umbral/curvebn.py

```diff
@@ -20,29 +21,27 @@ def __init__(self, bignum, curve_nid, group, order):
         self.order = order
 
     @classmethod
-    def gen_rand(cls, curve: ec.EllipticCurve = None):
+    def get_size(cls, curve: ec.EllipticCurve=None):
         """
-        Returns a BigNum object with a cryptographically secure BigNum based
-        on the given curve.
+        Returns the size (in bytes) of a CurveBN given the curve.
+        If no curve is provided, it uses the default.
         """
         curve = curve if curve is not None else default_curve()
-        curve_nid = backend._elliptic_curve_to_nid(curve)
-
-        group = backend._lib.EC_GROUP_new_by_curve_name(curve_nid)
-        backend.openssl_assert(group != backend._ffi.NULL)
+        return get_curve_keysize_bytes(curve)
 
-        order = backend._lib.BN_new()
-        backend.openssl_assert(order != backend._ffi.NULL)
-        order = backend._ffi.gc(order, backend._lib.BN_clear_free)
-
-        with backend._tmp_bn_ctx() as bn_ctx:
-            res = backend._lib.EC_GROUP_get_order(group, order, bn_ctx)
-            backend.openssl_assert(res == 1)
+    @classmethod
+    def gen_rand(cls, curve: ec.EllipticCurve=None):
+        """
+        Returns a CurveBN object with a cryptographically secure OpenSSL BIGNUM
+        based on the given curve.
+        """
+        curve = curve if curve is not None else default_curve()
+        curve_nid = backend._elliptic_curve_to_nid(curve)
 
-        new_rand_bn = backend._lib.BN_new()
-        backend.openssl_assert(new_rand_bn != backend._ffi.NULL)
-        new_rand_bn = backend._ffi.gc(new_rand_bn, backend._lib.BN_clear_free)
+        group = openssl._get_ec_group_by_curve_nid(curve_nid)
+        order = openssl._get_ec_order_by_curve_nid(curve_nid)
 
+        new_rand_bn = openssl._get_new_BN()
         rand_res = backend._lib.BN_rand_range(new_rand_bn, order)
```

Comment:
I just noticed that this allows the BN to be 0, which we don't want (we want it to be in the interval `[1, q-1]`, for `q` the order of the curve). A better way would be to call `BN_rand_range(new_rand_bn, order_minus_1)` and then add 1 to the result. This ensures the resulting BN is in `[1, q-1]`

Reply:
`BN_rand_range` generates numbers where `0 <= RAND  < Order`.

So, it's still possible that a number can be generated as 0, but not as the actual order.